### DISCUSSION TITLE
Handling of fluctuating energy in systems with time series data

### DISCUSSION THREAD

**Discussion by stockh0lm** [oemof/tespy#530] [Q&A]
I'm stuck on a crucial implementation detail and would like your thoughts. The scenario I describe is from a research project I'm currently working on with a company. It is prototypical for the reoccurring problem of fluctuating regenerative energy sources, and I need a robust, elegant solution that I can build upon. 

The company aims to save costs (energy, maintenance, and investment costs). They consume a lot of electricity and have limited opportunities to reduce consumption. However, they could also use their waste heat as heating energy for the neighboring companies and thus save costs on balance.

The company's landlord wants to decarbonize. They currently heat with gas. Heating with the company’s waste heat would significantly reduce the landlord’s CO2 footprint. They want to cooperate. 

The company has several larger ovens and an exhaust air and recooling system. Management is considering installing a heat recovery system and determining how to run it. They record electricity data in minute intervals, but temperature or humidity values still need to be recorded. However, I have a series of temperature and humidity measurements that they recorded manually. In the long term, continuous measurements of temperature, mass flow, and humidity of the exhaust air and pressure, mass flow, and temperature in the water circuit are used at various points to build a practical digital twin.

I am considering how best to manage my energy management system's temperatures/heat energy. I know the bus system that Francesc sometimes used with TesPy: A system could have several energy buses, each with a fixed set of properties, such as “electric power, 110kV, 2000A” or “superheated steam, 200°C, 20 bar”. That would be a constant power that this bus can deliver. Then there are components such as pumps, valves, heat pumps, or (in the opposite direction) turbines and generators that connect different buses and convert energy between them. In my case, one bus might be waste heat at room temperature, another bus the local heating flow temperature (approx. 80°C), and the connecting component would be a heat pump using electrical power to move heat between those two buses.

But what about fluctuating temperatures and outputs? Fluctuations are something that should not exist with buses. An alternative to the bus system would be energy packages that move between components at fixed intervals. Depending on the process section, this could be waste heat from an oven (with the properties temperature, pressure, water content) or hot water, pressure, water content) or hot water in a layer in the buffer tank (water, pressure, temperature). I would not handle the internal states of components in detail; I would only abstract their input and output values via linearized characteristic lines. Buses and energy package systems deal with power or energy transport or balancing.

My specific system at the company above has three temperature levels. At least one fluctuates wildly, which would contradict treatment as a bus.

1. Room temperature: medium: Air, 20-30°C, approx. 40% humidity. The temperature rises over approximately 4-8 hours and then slowly falls again, depending on whether the ovens are running. The ovens often do not run for a few days, e.g., at weekends. This temperature level does not fulfill the requirement of a bus. Perhaps it could be called a variable or fluctuating bus.
2. The flow of the local heating is water, 80°C, as constant as possible.
3. Electricity: Another energy bus is electric power: 400V 32A (three-phase current)

So, I have two constant buses and a third, fluctuating, variable bus that may be inactive for extended periods. How can I solve this elegantly in terms of programming? I could create energy packets per time increment (15 minutes), which can fluctuate in energy content and parameters. The packets would be constant over time with the constant buses but implemented similarly for consistency.

It is unclear how much energy a heat-energy package contains: using the example of the 20m³/30°C/40% humidity energy packet, it depends on the temperature to which I can cool it down and how much energy I can extract from it. I would have to calculate thermodynamically to handle the separate energy packages correctly. It's possible to do that in Linear Programming (which I plan to use to minimize the costs of the system and the energy requirements) with multidimensional piecewise linear approximation.

- Temperature Management: Given that the room temperature fluctuates significantly, how can I effectively manage these variations within my energy management system? Can a fluctuating or variable "bus" concept be integrated to handle this irregularity?

- Energy Packet Calculation: How should I calculate and manage energy packets for varying temperatures and outputs over time? What would be an efficient way to represent these fluctuations programmatically?

- Thermodynamic Calculations: Given the complexity of fluctuating heat and moisture levels, what method should I use to perform accurate thermodynamic calculations for each energy packet? How can I incorporate these calculations into my system design to ensure energy efficiency and cost-effectiveness? Piecewise linear approximations are feasible but blow up the number of decision variables with linear programming.

**Comment by fwitte** [comment]
Hi @stockh0lm and thank you for reaching out. This is quite extensive, I will respond with a little bit more time... :).

Best!
